### PR TITLE
Use SSCG for creation of certificates

### PR DESCRIPTION
**Pull request by sgallagher**
Last year, I wrote a [blog post](https://sgallagh.wordpress.com/2016/05/02/self-signed-ssltls-certificates-why-they-are-terrible-and-a-better-alternative/) on some of the reasons why self-signed certificates are bad and what we can do to be safer in those environments where their use (or the use of something like them) is unavoidable.

tl;dr version: We create a single-use Certificate Authority with X509v3 restrictions to allow it to sign exactly one certificate, then destroy the private key so it can never be reused.

To that end, I built a tool called SSCG (Simple Signed Certificate Generator) to offer an easy way to generate these safer certificates. This pull request adds support to Cockpit to use the sscg when it needs to generate a certificate (because none is currently present).

The first commit adds all of the logic necessary to fork a request to sscg and generate the certificates. It is also designed in such a way that if sscg does not exist (or fails for some reason), it will fall back to the existing OpenSSL approach for generating self-signed certificates.

The second commit modifies the RPM spec file to build with SSCG support by default and to add Recommends: to it on those platforms where SSCG is available.

### CODE REVIEW

**Comment by sgallagher** [comment]
I modified the patches after an in-person conversation with @stefwalter 

He advised me to remove the configure flag for using SSCG since it is intended to gracefully fail back to openssl at runtime anyway.

**Comment by dperpeet** [comment]
Do we need to update the guide for this? http://cockpit-project.org/guide/latest/https.html#https-certificates

**Comment by sgallagher** [comment]
> Do we need to update the guide for this? http://cockpit-project.org/guide/latest/https.html#https-certificates

Yeah, good idea. I've pushed a patch that includes that.

**Comment by stefwalter** [comment]
This adds a new dependency. I would wait on this until we stop pushing to Fedora 25.

**Comment by sgallagher** [comment]
@stefwalter Sorry, I don't understand your specific concern. It adds a dependency on a component that exists in F25 and is the same version as F26.

That said, if you prefer, I can modify the spec to do the `Recommends: sscg` only on F26+

Or I can bring back the configure option to allow F25 to not attempt to use `sscg`.

**Comment by petervo** [comment]
@sgallagher I think we are good to merge this now, but needs a rebase.

**Comment by sgallagher** [comment]
@dperpeet 
I rebased atop master and re-tested on Fedora 26, both with `sscg` available and unavailable on the system.

File: src/ws/cockpitcertificate.c [hunk 1]
```diff
@@ -32,15 +32,45 @@
...
 static gchar *
-generate_subject (void)
+get_common_name (void)
+{
+  int ret;
+  gchar *hostname;
+  gchar *cn;
+
+  hostname = g_malloc(HOST_NAME_MAX + 1);
+  if (!hostname) return NULL;
```
Comment (petervo) [inline src/ws/cockpitcertificate.c]: We try not to have single line if's

File: src/ws/cockpitcertificate.c [hunk 2]
```diff
@@ -32,15 +32,45 @@
...
   static const char HEX[] = "0123456789abcdef";
-  gchar hostname[HOST_NAME_MAX + 1] = { 0, };
   gchar *content;
-  gchar *subject;
-  gchar *cn;
-  int ret;
+  gchar *machine_id = NULL;
 
+  if (g_file_get_contents ("/etc/machine-id", &content, NULL, NULL))
+    {
```
Comment (petervo) [inline src/ws/cockpitcertificate.c]: Brackets are unnecessary

File: src/ws/cockpitcertificate.c [hunk 3]
```diff
@@ -32,15 +32,45 @@
...
 
 static gchar *
-generate_subject (void)
+get_common_name (void)
+{
+  int ret;
+  gchar *hostname;
+  gchar *cn;
+
+  hostname = g_malloc(HOST_NAME_MAX + 1);
```
Comment (petervo) [inline src/ws/cockpitcertificate.c]: space before (

File: src/ws/cockpitcertificate.c [hunk 4]
```diff
@@ -32,15 +32,45 @@
...
+  if (!hostname) return NULL;
+
+  ret = gethostname (hostname, HOST_NAME_MAX);
+  if (ret < 0 || g_str_equal (hostname, ""))
+    cn = g_strdup("localhost");
```
Comment (petervo) [inline src/ws/cockpitcertificate.c]: space before (

File: src/ws/cockpitcertificate.c [hunk 5]
```diff
@@ -32,15 +32,45 @@
...
+  else
+    cn = g_strdup(hostname);
```
Comment (petervo) [inline src/ws/cockpitcertificate.c]: space before (

File: src/ws/cockpitcertificate.c [hunk 6]
```diff
@@ -54,23 +84,21 @@ generate_subject (void)
    *
    */
 
-  ret = gethostname (hostname, sizeof (hostname));
-  if (ret < 0 || g_str_equal (hostname, ""))
-    cn = "localhost";
-  else
-    cn = hostname;
+  cn = get_common_name();
```
Comment (petervo) [inline src/ws/cockpitcertificate.c]: space before (

File: src/ws/cockpitcertificate.c [hunk 7]
```diff
@@ -54,23 +84,21 @@ generate_subject (void)
...
-                                 g_strstrip (g_strcanon (content, HEX, ' ')), cn);
-      g_free (content);
+                                 machine_id, cn);
     }
   else
     {
       subject = g_strdup_printf ("/CN=%s", cn);
     }
 
+  g_free(cn);
```
Comment (petervo) [inline src/ws/cockpitcertificate.c]: space before (

File: src/ws/cockpitcertificate.c [hunk 8]
```diff
@@ -54,23 +84,21 @@ generate_subject (void)
...
+  g_free(machine_id);
```
Comment (petervo) [inline src/ws/cockpitcertificate.c]: space before (

File: src/ws/cockpitcertificate.c [hunk 9]
```diff
@@ -120,6 +148,65 @@ openssl_make_dummy_cert (const gchar *key_file,
...
+sscg_make_dummy_cert (const gchar *key_file,
+                      const gchar *cert_file,
+                      const gchar *ca_file,
+                      GError **error)
+{
+  gboolean ret = FALSE;
+  gint exit_status;
+  gchar *stderr_str = NULL;
+  gchar *command_line = NULL;
+  gchar *cn = get_common_name();
```
Comment (petervo) [inline src/ws/cockpitcertificate.c]: space before (

File: src/ws/cockpitcertificate.c [hunk 10]
```diff
@@ -120,6 +148,65 @@ openssl_make_dummy_cert (const gchar *key_file,
...
+  gchar *machine_id = get_machine_id();
```
Comment (petervo) [inline src/ws/cockpitcertificate.c]: space before (

File: src/ws/cockpitcertificate.c [hunk 11]
```diff
@@ -120,6 +148,65 @@ openssl_make_dummy_cert (const gchar *key_file,
...
+  gchar *org;
+
+  if (machine_id)
+    {
```
Comment (petervo) [inline src/ws/cockpitcertificate.c]: Brackets are unnecessary

File: src/ws/cockpitcertificate.c [hunk 12]
```diff
@@ -120,6 +148,65 @@ openssl_make_dummy_cert (const gchar *key_file,
...
+      org = machine_id;
+    }
+  else
+    {
```
Comment (petervo) [inline src/ws/cockpitcertificate.c]: Brackets are unnecessary

File: src/ws/cockpitcertificate.c [hunk 13]
```diff
@@ -120,6 +148,65 @@ openssl_make_dummy_cert (const gchar *key_file,
...
+  g_info ("Generating temporary certificate using: %s", command_line);
+
+  if (!g_spawn_sync (NULL, (gchar **)argv, NULL, G_SPAWN_SEARCH_PATH, NULL, NULL,
+                     NULL, &stderr_str, &exit_status, error) ||
+      !g_spawn_check_exit_status (exit_status, error))
+    {
+      /* Failure of SSCG is non-fatal */
+      g_info ("Error generating temporary dummy cert using sscg, "
+              "falling back to openssl");
+      g_clear_error(error);
```
Comment (petervo) [inline src/ws/cockpitcertificate.c]: space before (

File: src/ws/cockpitcertificate.c [hunk 14]
```diff
@@ -175,12 +266,25 @@ generate_temp_cert (const gchar *dir,
       goto out;
     }
 
+  /* First, try to create a private CA and certificate using SSCG */
+  if (sscg_make_dummy_cert(cert_path, cert_path, ca_path, error))
```
Comment (petervo) [inline src/ws/cockpitcertificate.c]: space before (

**Comment by sgallagher** [comment]
@petervo Yes, please clean those up on merge, and thanks!

**Pull request closed without merging**